### PR TITLE
 Tx verification in DemoApp

### PR DESCRIPTION
**Pull request by bkolad**
This PR introduces `TxVerifier` trait. More refined implementation of `TxVerification` and better docs will come in the following PRs.

### CODE REVIEW

File: demo-app/src/data_generation.rs [hunk 1]
```diff
@@ -88,3 +105,18 @@ impl QueryGenerator {
         Runtime::<MockContext>::encode_value_setter_query(query_message)
     }
 }
+
+impl Transaction<MockContext> {
+    pub fn new(msg: Vec<u8>, pub_key: MockPublicKey) -> Self {
+        let _hash = <MockContext as Spec>::Hasher::hash(&msg);
```
Comment (preston-evans98) [inline demo-app/src/data_generation.rs:111]: (Question): Why do we create and drop the message hash here?

File: demo-app/src/main.rs [hunk 2]
```diff
@@ -1,27 +1,27 @@
+mod batch;
 mod data_generation;
 mod helpers;
 mod runtime;
-mod types;
+mod sft;
```
Comment (preston-evans98) [inline demo-app/src/main.rs:5]: nit: `sft` -> `stf`

File: demo-app/src/tx_verifier.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    pub(crate) runtime_msg: Vec<u8>,
+    pub(crate) _nonce: u64,
+}
+
+/// TxVerifier encapsulates Transaction verification.
+pub(crate) trait TxVerifier {
+    type Context: Context;
+
+    /// Runs stateless checks against RawTxs.
+    fn verify_txs_stateless(
```
Comment (preston-evans98) [inline demo-app/src/tx_verifier.rs:34]: Should add a version of this method (`verify_tx_stateless`) that works on a single tx? I think that will be useful for sequencers. We can keep the batch method too if we want, and just add a default impl that calls `verify_tx_stateless` in a loop

**Review by preston-evans98 — COMMENTED** [review]
One overall comment: it makes code review much easier if we separate out refactoring changes from functional changes. I think it's fine to keep them in the same PR, but having separate commits is nice - and separate PRs is even nicer.

File: sov-modules/sov-modules-api/src/lib.rs [hunk 4]
```diff
@@ -16,6 +16,25 @@ pub use response::{CallResponse, QueryResponse};
...
+#[derive(Error, Debug)]
+pub enum SigVerificationError {
+    #[error("Bad signature")]
+    BadSignature,
+}
+
+pub trait Signature {
+    type PublicKey;
+
+    fn verify(
```
[the remark this answers is not in the document]
Comment (bkolad) [reply sov-modules/sov-modules-api/src/lib.rs:31]: Added an issue: https://github.com/Sovereign-Labs/sovereign/issues/130

File: demo-app/src/data_generation.rs (hunk 1 above)
Comment (bkolad) [reply demo-app/src/data_generation.rs]: This was a leftover, removed.

File: demo-app/src/main.rs (hunk 2 above)
Comment (bkolad) [reply demo-app/src/main.rs]: done

File: demo-app/src/tx_verifier.rs (hunk 3 above)
Comment (bkolad) [reply demo-app/src/tx_verifier.rs:38]: done

File: demo-app/src/tx_verifier.rs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+pub struct Transaction<C: sov_modules_api::Context> {
+    pub signature: C::Signature,
+    pub pub_key: C::PublicKey,
+    pub runtime_msg: Vec<u8>,
+    pub nonce: u64,
+}
+
+/// VerifiedTx is a Transaction after verification.
+pub(crate) struct VerifiedTx<C: Context> {
+    pub(crate) sender: C::PublicKey,
```
[the remark this answers is not in the document]
Comment (preston-evans98) [reply demo-app/src/tx_verifier.rs:24]: Sounds good

File: sov-modules/sov-modules-api/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (preston-evans98) [reply sov-modules/sov-modules-api/src/lib.rs:31]: Thanks for making the issue!

**Review by preston-evans98 — APPROVED** [review]

**Pull request merged**